### PR TITLE
mbf21: things

### PR DESCRIPTION
**Pull request by rfomin**
See: https://github.com/fabiangreffrath/woof/blob/56960cf7afd83443e93db0f6a822cc7e179de549/docs/mbf21.md#things

### CODE REVIEW

File: Source/d_deh.c [hunk 1]
```diff
@@ -1086,6 +1098,28 @@ struct {
   {"TRANSLUCENT",  0x80000000}, // apply translucency to sprite (BOOM)
 };
 
+#define DEH_MOBJFLAGMAX_MBF21 (sizeof(deh_mobjflags_mbf21) / sizeof(*deh_mobjflags_mbf21))
+
+static const struct deh_flag_s deh_mobjflags_mbf21[] = {
```
Comment (rfomin) [inline Source/d_deh.c:1103]: MBF21 added new mobj flags, so we need to add a new field to mobj_t and broke saves. Maybe we should introduce some kind of versioning?

File: Source/p_mobj.h [hunk 2]
```diff
@@ -355,6 +376,9 @@ typedef struct mobj_s
     fixed_t		oldy;
     fixed_t		oldz;
     angle_t		oldangle;
+
+    // mbf21
+    int flags2;
```
Comment (rfomin) [inline Source/p_mobj.h:381]: MBF21 added new mobj flags, so we need to add a new field to `mobj_t` and broke saves. Maybe we should introduce some kind of versioning?

Comment (fabiangreffrath) [reply Source/p_mobj.h:381]: Yes, we should replace `VERSIONID` with a versioned Woof-specific identifier each time we break savegame compatibility.

Comment (fabiangreffrath) [reply Source/p_mobj.h:381]: We aren't MBF-compatible anymore for quite some time anyway.

Comment (rfomin) [reply Source/p_mobj.h:381]: Should we try to support the previous version of saves?

Comment (fabiangreffrath) [reply Source/p_mobj.h:381]: Yes, I think this would be only fair.

File: Source/g_game.c [hunk 3]
```diff
@@ -1540,7 +1540,8 @@ static void G_DoSaveGame(void)
   memset (name2,0,sizeof(name2));
 
   // killough 2/22/98: "proprietary" version string :-)
-  sprintf (name2,VERSIONID,MBFVERSION);
+  //sprintf (name2,VERSIONID,MBFVERSION);
+  sprintf (name2,VERSIONID,MBF21VERSION);
```
Comment (rfomin) [inline Source/g_game.c:1544]: After a quick look at the `save_g.c` file, I think compatibility with old saves is difficult. My first idea is rewriting most of it in the style of Chocolate Doom. Do you have ideas?

Comment (fabiangreffrath) [reply Source/g_game.c:1544]: That will be a huge amount of work, dissecting all the structs into their individual members. The advantage, though, would be that we could support both 32 and 64 bit saves and even old saves from MBF.exe.

Comment (rfomin) [reply Source/g_game.c:1544]: I just don't see any other way. Even I have saves and I don't want to break them :)

File: Source/m_random.h [hunk 4]
```diff
@@ -118,6 +118,9 @@ typedef enum {
   pr_dropoff,                 // #60
   pr_randomjump,              // #61
   pr_defect,                  // #62
+  // mbf21
+  pr_heretic,                 // #63
```
Comment (kraflab) [inline Source/m_random.h:122]: I don't think the index in this list matters, so you could just as well skip `pr_heretic` if you want. MBF21 doesn't support splitting the rng between all these entries - in practice there is only one pool because demo_insurance is always 0. I added `pr_heretic` to dsda-doom for example purely as a self-documenting marker since it didn't matter one way or the other.

Comment (rfomin) [reply Source/m_random.h:122]: Good to know, thanks.

File: Source/g_game.c (hunk 3 above)
Comment (fabiangreffrath) [reply Source/g_game.c:1544]: > I just don't see any other way. Even I have saves and I don't want to break them :)

I see.

> Or port Woof to Choco/Crispy as another Doom engine. Maybe it's even easier.

Not in this case. The main advantage would be shared audio/video/input code - but I think we got this quite right already. The savegame code would have to be rewritten one way or the other.

Comment (rfomin) [reply Source/g_game.c:1544]: > The main advantage would be shared audio/video/input code

Also netcode, much better dehacked and opl emulation. Sometimes I seriously think about it.

**Comment by fabiangreffrath** [comment]
Is this ready for merging?

**Comment by rfomin** [comment]
Yes, but now we have a new version of the saves. I'm starting to rewrite `p_saveg.c`

File: Source/p_map.c [hunk 5]
```diff
@@ -470,6 +470,26 @@ static boolean PIT_CheckLine(line_t *ld) // killough 3/26/98: make static
...
+        source->type == target->type
+      ) ||
+      ( // target type has special behaviour, and things have the same group
+        mobjinfo[target->type].projectile_group != PG_DEFAULT &&
+        mobjinfo[target->type].projectile_group == mobjinfo[source->type].projectile_group
+      )
+    );
+}
+
+
```
Comment (fabiangreffrath) [inline Source/p_map.c:492]: Excess blank line.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Some changes are a bit hard to parse, e.g. the one in `G_SetFastParms()`, but I guess it's alright. Could you please add comments to all added code blocks to indicate they are part of MBF21? Thanks!

**Comment by kraflab** [comment]
We added `meleerange` in the scope of one of the last codepointers, fyi: https://github.com/kraflab/dsda-doom/pull/46/files

**Comment by fabiangreffrath** [comment]
@rfomin Are you going to add the variable `meleerange` implementation to this PR or should I apply this first and everything else comes in separate PRs?

**Comment by rfomin** [comment]
> Are you going to add the variable `meleerange` implementation

Yes, I plan to add it. I was distracted by savegame code.
